Python/pywarpx/diagnostics/data/BaseReader.py

```diff
@@ -0,0 +1,173 @@
+
+class DataReader():
+    """
+    Base class that all data readers should inherit from.
+    Works only in Cartesian geometry.
+    """
+
+    def __init__(self, run_directory, file_prefix=None):
+        """
+        Parameters
+        ----------
+        run_directory: string
+            Path to the run directory of WarpX.
+        """
+        if run_directory is None:
+            raise ValueError('The run_directory parameter can not be None!')
+
+        self.run_directory = run_directory
+        self.data_file_prefix = None
+        self.data_file_suffix = None
+
+        self.has_species = None
+
+    def get_data_path(self, subdir='reducedfiles'):
+        """
+        Returns the path to the data file.
+        Parameters
+        ----------
+        subdir: string
+            Name of reducedfiles directory, default is 'reducedfiles'.
+        Returns
+        -------
+        A string with a path to the current diagnostic (assumed unique).
+        """
+
+        sim_output_dir = os.path.join(self.run_directory, 'diags/'+subdir+'/')
+        if not os.path.isdir(sim_output_dir):
+            raise IOError('The simulation directory does not exist inside '
+                          'path:\n  {}\n'
+                          'Did you set the proper path to the run directory?\n'
+                          'Did the simulation already run?'
+                          .format(self.run_directory))
+
+        data_file_path = os.path.join(
+            sim_output_dir,
+            self.data_file_prefix +
+            self.data_file_suffix
+        )
+        if not os.path.isfile(data_file_path):
+            raise IOError('The file {} does not exist.\n'
+                          'Did the simulation already run?'
+                          .format(data_file_path))
+
+        return data_file_path
+
+    def get_ncols(self):
+        ''' Returns the number of columns of the reduced diagnostic.'''
+        data_file_path = self.get_data_path()
+        data = np.loadtxt(data_file_path)
+        if data.ndim > 1:
+            ncols = np.shape(data)[1]
+        else:
+            ncols = len(data)
+        return ncols
+
+    def get_nspecies(self, subtract=None, divide=None):
+        '''Returns the number of species involved in the diagnostic.'''
+        # deduces number of species from number of columns
+        if self.has_species == True:
+            ncols = self.get_ncols()
+            # remove <substract> columns and divide to avoid counting multiple entries
+            nspecies = int((ncols - subtract)/divide)
+            return nspecies
+        else:
+            return 0
+
+    def get_species_names(self, string=None, start=None, step=None):
+        '''Returns the names of the species involved in the diagnostic.'''
+        # deduces names of species from the header by considering only the relevant entries
+        # finds the string by locating the units
+        if self.has_species == True:
+            species_names = []
```

Comment:
Doesn't pandas read in the header? Can you get the species names from there?

Reply:
Yes, I started using pandas after doing this. I will fix it.